lib/ansible/module_utils/keycloak.py

```diff
@@ -339,3 +343,164 @@ def delete_client_template(self, id, realm="master"):
+        except Exception as e:
+            self.module.fail_json(
+                msg='Could not obtain user %s for realm %s: %s'
+                    % (id, realm, str(e)))
+
+    def get_user_id(self, name, realm='master'):
+        """ Obtain user id by name
+
+        :param name: name of user to be queried
+        :param realm: client template from this realm
+        :return: user id (usually a UUID)
+        """
+        result = self.get_user_by_name(name, realm)
+        if isinstance(result, dict) and 'id' in result:
+            return result['id']
+        else:
+            return None
+        
+    def get_user_by_name(self, name, realm='master'):
+        """ Obtain user representation by name
+
+        :param name: name of user to be queried
+        :param realm: user from this realm
+        :return: dict of user representation or None if none matching exist
+        """
+        result = self.get_users(realm)
+        if isinstance(result, list):
+            result = [x for x in result if x['username'] == name]
+            if len(result) > 0:
+                return result[0]
+        return None
+    
+    def create_user(self, user_representation, realm="master"):
+        """ Create a user in keycloak
+
+        :param user_representation: user representation of user to be created. Must at least contain field userId
+        :param realm: realm for user to be created
+        :return: HTTPResponse object on success
+        """
+        # Keycloak wait username as key for the keycloak user username.
+        try:
+            user_name = user_representation.pop('keycloakUsername')
+        except KeyError:
+            self.module.fail_json(
+                msg='User name needs to be specified when creating a new user',
+                user_representation=user_representation
+            )
+        else:
+            user_representation.update({'username': user_name})
+        user_url = URL_USERS.format(url=self.baseurl, realm=realm)
+
+        try:
+            return open_url(user_url, method='POST', headers=self.restheaders,
+                            data=json.dumps(user_representation), validate_certs=self.validate_certs)
+        except Exception as e:
+            self.module.fail_json(msg='Could not create user %s in realm %s: %s'
+                                      % (user_representation['username'], realm, str(e)),
+                                  payload=user_representation)
+            
+    def update_user(self, uuid, user_representation, realm="master"):
+        """ Update an existing user
+        :param uuid: id of user to be updated in Keycloak
+        :param user_representation: corresponding (partial/full) user representation with updates
+        :param realm: realm the user is in
+        :return: HTTPResponse object on success
+        """
+        # Keycloak wait username as key for the keycloak user username.
+        try:
+            user_representation.pop('keycloakUsername')
```

Comment:
Keycloak API returns error 409 Conflict if a username is sent (even if it is the same than the existing one).

The module does not allow you to change a username, because you cannot send to the module a id and username. But, when you specified a user to update with the username, it stays in the `user_representation` dictionary and thus should be deleted.

I can improve the comment about it.

Reply:
ok